lsp-types/src/Language/LSP/Types/Common.hs

```diff
@@ -10,14 +13,41 @@ module Language.LSP.Types.Common (
   , List (..)
   , Empty (..)
   , Int32
-  , Word32 ) where
+  , UInt ) where
 
 import Control.Applicative
 import Control.DeepSeq
 import Data.Aeson
+import Data.Coerce (coerce)
+import Data.Euclidean as E
 import Data.Int (Int32)
-import Data.Word (Word32)
-import GHC.Generics
+import Data.Mod.Word
+import Text.Read (Read(readPrec))
+import GHC.Generics hiding (UInt)
+
+-- | The "uinteger" type in the LSP spec.
+--
+-- Unusually, this is a **31**-bit unsigned integer, not a 32-bit one.
+newtype UInt = UInt (Mod 31)
+  deriving newtype (Num, Bounded, Enum, Eq, Ord, Show)
+  deriving stock (Generic)
+  deriving anyclass (NFData)
+
+instance Read UInt where
+  readPrec = fromInteger <$> readPrec
+
+instance Real UInt where
+  toRational (UInt u) = toRational $ unMod u
+
+instance Integral UInt where
+  quotRem (UInt x) (UInt y) = coerce $ E.quotRem x y
```

Comment:
```suggestion
  quotRem (UInt x) (UInt y) = coerce $ fromIntegral $ quotRem (unMod x) (unMod y)
```
`E.quotRem` (division in a modular ring) is not quite what you want here.

Reply:
Hmm, I mistakenly thought it would be equivalent, thank you.